### ISSUE TITLE
Error when running update

### ISSUE THREAD

**Issue by lpoorvin** [ncbi/pgap#7]
I'm trying to update a previously working 2018-11-07.build3190 stand-alone install to the current 2019-02-11.buiild3477 on Linux Mint 18 (based on Ubuntu 16.04 LTS).  Following the instructions in the quick start guide, I grabbed the current version via curl, set the execute bit and ran 
`$ ./pgap.py --update`
Needed updates appeared to download successfully, but the install eventually failed with a type error; relevant output is below.  Trying to rerun the update returns the same error.  Am I just missing something obvious? Thanks!

```
$ ./pgap.py --update
Updating PGAP to version 2019-02-11.build3477 (previous version was 2018-11-07.build3190)
Downloading (as needed) PGAP Docker image version 2019-02-11.build3477
2019-02-11.build3477: Pulling from ncbi/pgap
af4b0a2388c6: Already exists 
...
64f7899a4cdc: Pull complete 
Digest: sha256:d8e38a6678fc64507472cdfe02c45d12bffb24ee43136a0f89a8144e6082ede8
Status: Downloaded newer image for ncbi/pgap:2019-02-11.build3477
Downloading PGAP reference data version 2019-02-11.build3477
Downloaded 14045155161 of 14045155161 bytes (100.00%)
Traceback (most recent call last):
  File "./pgap.py", line 282, in <module>
    main()
  File "./pgap.py", line 268, in main
    version = setup(args.update, args.local_runner)
  File "./pgap.py", line 178, in setup
    f.write('{}\n'.format(latest))
TypeError: write() argument 1 must be unicode, not str
```

**Comment by slottad** [comment]
Sorry for the difficulties. I have committed some changes. Please update your copy and try again.

**Comment by lpoorvin** [comment]
Thanks for the quick reply.  Updating now worked.  However, trying to annotate a custom genome failed.

```
$ ./pgap.py test_genomes/A88_pilon/A88_pilon.yaml 
Traceback (most recent call last):
  File "./pgap.py", line 282, in <module>
    main()
  File "./pgap.py", line 279, in main
    run(version, input, args.output, debug)
  File "./pgap.py", line 199, in run
    f.write('\n')
TypeError: write() argument 1 must be unicode, not str
```
I noted that the last commit had a comment re: python 2/3 compatability so I tried running in a conda environment with Python 3.7 (default was 2.7).  This time the annotation started, but failed a few minutes in.

```
$ ./pgap.py test_genomes/A88_pilon/A88_pilon.yaml 
/usr/bin/cwltool 1.0.20181217162649
Resolved 'pgap.cwl' to 'file:///pgap/pgap.cwl'
/usr/lib64/python3.6/site-packages/ruamel/yaml/constructor.py:277: DuplicateKeyFutureWarning: while constructing a mapping
  in "file:///pgap/user_input/pgap_input.yaml", line 1, column 1
_strain/my_strain.yaml
...
[job Prepare_Unannotated_Sequences] Max memory used: 61MiB
[job Prepare_Unannotated_Sequences] Job error:
Error collecting output for parameter 'master_desc':
bacterial_prepare_unannotated.cwl:45:7: Did not find output file with glob pattern: '['master-desc.asn']'
[job Prepare_Unannotated_Sequences] completed permanentFail
[step Prepare_Unannotated_Sequences] Output is missing expected field file:///pgap/wf_common.cwl#Prepare_Unannotated_Sequences/master_desc
[step Prepare_Unannotated_Sequences] Output is missing expected field file:///pgap/wf_common.cwl#Prepare_Unannotated_Sequences/sequences
[step Prepare_Unannotated_Sequences] completed permanentFail
[workflow standard_pgap] completed permanentFail
[step standard_pgap] completed permanentFail
[workflow ] completed permanentFail
Final process status is permanentFail
{
    "gbent": null,
    "gbk": null,
    "gff": null,
    "input_fasta": {
        "class": "File",
        "location": "file:///pgap/output/genome.fa",
        "size": 7957297,
        "basename": "genome.fa",
        "checksum": "sha1$c5f0c5156b1091c112c93d3498f421ecd6502db8",
        "path": "/pgap/output/genome.fa"
    },
    "input_submol": {
        "class": "File",
        "location": "file:///pgap/output/submol.yaml",
        "size": 807,
        "basename": "submol.yaml",
        "checksum": "sha1$ed736f025d23d959e327146ee6c48d73a2080eda",
        "path": "/pgap/output/submol.yaml"
    },
    "nucleotide_fasta": null,
    "protein_fasta": null
}
Traceback (most recent call last):
  File "./pgap.py", line 282, in <module>
    main()
  File "./pgap.py", line 279, in main
    run(version, input, args.output, debug)
  File "./pgap.py", line 233, in run
    subprocess.check_call(cmd)
  File "/home/ngs/anaconda2/envs/pgap/lib/python3.7/subprocess.py", line 347, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'run', '-i', '--user', '1000:1000', '--volume', '/home/ngs/pgap/input-2019-02-11.build3477:/pgap/input:ro', '--volume', '/home/ngs/pgap/test_genomes/A88_pilon:/pgap/user_input', '--volume', '/home/ngs/pgap/output/pgap_input.yaml:/pgap/user_input/pgap_input.yaml:ro', '--volume', '/home/ngs/pgap/output:/pgap/output:rw', '--volume', '/home/ngs/pgap/output/log:/log/srv', 'ncbi/pgap:2019-02-11.build3477', 'cwltool', '--outdir', '/pgap/output', 'pgap.cwl', '/pgap/user_input/pgap_input.yaml']' returned non-zero exit status 1.
```

**Comment by slottad** [comment]
The pgap.py merely starts the pipeline, which runs entirely under docker. At that point, the python environment under which pgap.py ran is moot. So now you have a different problem.

Is it possible we could see your input files?

**Comment by lpoorvin** [comment]
Sorry, the second error was my fault, I've updated the comment to reflect that.  The annotation appears to be running, however, I'm getting a string of these warnings:

`Use of assignment to $[ is deprecated at /panfs/pan1.be-md.ncbi.nlm.nih.gov/gpipe/ThirdParty/tRNAscan-SE/production/arch/x86_64/bin/tRNAscan-SE line 3562.`

**Comment by slottad** [comment]
I fixed the first error, please try your custom genome again.

Those warnings are known, irritating, and on our list. For now, you should not worry about them.

**Comment by lpoorvin** [comment]
OK, it seems to be running fine under py2.7 with the latest update.  Thanks!

**Comment by lpoorvin** [comment]
The annotation ran for a while but then errored out again.  Any ideas?

```
./pgap.py test_genomes/A88_pilon/A88_pilon.yaml 
/usr/bin/cwltool 1.0.20181217162649
Resolved 'pgap.cwl' to 'file:///pgap/pgap.cwl'
/usr/lib64/python3.6/site-packages/ruamel/yaml/constructor.py:277: DuplicateKeyFutureWarning: while constructing a mapping
  in "file:///pgap/user_input/pgap_input.yaml", line 1, column 1
found duplicate key "supplemental_data" with value "ordereddict()" (original value: "ordereddict()")
  in "file:///pgap/user_input/pgap_input.yaml", line 13, column 1

...

[workflow standard_pgap] starting step Prepare_Unannotated_Sequences
[step Prepare_Unannotated_Sequences] start
[job Prepare_Unannotated_Sequences] /tmp/r11171ka$ bacterial_prepare_unannotated \
    -submission-mode-genbank \
    -asn-cache \
    /tmp/tmpjf4ey5se/stgdfd624e2-d9a6-4c6c-8dd4-6bb41a3f104b/sequence_cache \
    -gc-assembly \
    /tmp/tmpjf4ey5se/stg9aaabddc-c012-45d4-bd27-5c04613f4df1/gencoll.asn \
    -ids \
    /tmp/tmpjf4ey5se/stgdd21aed1-e7d0-4f45-972b-0e8527e61b34/ref_no_organelle.gi \
    -master-desc \
    master-desc.asn \
    -o \
    sequences.asn \
    -submit-block \
    /tmp/tmpjf4ey5se/stg9ad8bb17-773e-4c5d-9e20-5847f54c780a/submit_block_template.asn \
    -taxon-db \
    /tmp/tmpjf4ey5se/stge3204fe1-202e-45bd-b9a5-c353f8f418e4/taxonomy.sqlite3
[job Prepare_Unannotated_Sequences] Max memory used: 65MiB
[job Prepare_Unannotated_Sequences] Job error:
Error collecting output for parameter 'master_desc':
bacterial_prepare_unannotated.cwl:45:7: Did not find output file with glob pattern: '['master-desc.asn']'
[job Prepare_Unannotated_Sequences] completed permanentFail
[step Prepare_Unannotated_Sequences] Output is missing expected field file:///pgap/wf_common.cwl#Prepare_Unannotated_Sequences/master_desc
[step Prepare_Unannotated_Sequences] Output is missing expected field file:///pgap/wf_common.cwl#Prepare_Unannotated_Sequences/sequences
[step Prepare_Unannotated_Sequences] completed permanentFail
[workflow standard_pgap] completed permanentFail
[step standard_pgap] completed permanentFail
[workflow ] completed permanentFail
Final process status is permanentFail
{
    "gbent": null,
    "gbk": null,
    "gff": null,
    "input_fasta": {
        "class": "File",
        "location": "file:///pgap/output/genome.fa",
        "size": 6654733,
        "basename": "genome.fa",
        "checksum": "sha1$5363718ce016688a664f9c4ed7ae2a5bc1a1a819",
        "path": "/pgap/output/genome.fa"
    },
    "input_submol": {
        "class": "File",
        "location": "file:///pgap/output/submol.yaml",
        "size": 807,
        "basename": "submol.yaml",
        "checksum": "sha1$ed736f025d23d959e327146ee6c48d73a2080eda",
        "path": "/pgap/output/submol.yaml"
    },
    "nucleotide_fasta": null,
    "protein_fasta": null
}
Traceback (most recent call last):
  File "./pgap.py", line 275, in <module>
    main()
  File "./pgap.py", line 272, in main
    run(version, input, args.output, debug)
  File "./pgap.py", line 226, in run
    subprocess.check_call(cmd)
  File "/home/ngs/anaconda2/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['docker', 'run', '-i', '--user', '1000:1000', '--volume', '/home/ngs/pgap/input-2019-02-11.build3477:/pgap/input:ro', '--volume', '/home/ngs/pgap/test_genomes/A88_pilon:/pgap/user_input', '--volume', '/home/ngs/pgap/output/pgap_input.yaml:/pgap/user_input/pgap_input.yaml:ro', '--volume', '/home/ngs/pgap/output:/pgap/output:rw', '--volume', '/home/ngs/pgap/output/log:/log/srv', 'ncbi/pgap:2019-02-11.build3477', 'cwltool', '--outdir', '/pgap/output', 'pgap.cwl', '/pgap/user_input/pgap_input.yaml']' returned non-zero exit status 1
```

**Comment by azat-badretdin** [comment]
Something must have happened early in execution of application `bacterial_prepare_unannotated`. Absence of any logging from that app indicates to that. I identified the code range where this might happen, but nothing clicks for now: that part is trivial. 

I wonder if it is possible to obtain input files or you can re-create the error using a public genome.

**Comment by lpoorvin** [comment]
I think I had a problem in my input.yaml file.  I rewrote it from scratch (I had been using a modified version from a previous run) and the annotation *seems* to be working.  A genome this size (~6.6 Mbp) takes 24+ hours to run on my system, I'll keep my fingers crossed and let you know how it shakes out.  Thanks for everyone's help!

**Comment by azat-badretdin** [comment]
Thanks for the feedback. 

>  A genome this size (~6.6 Mbp) takes 24+ hours to run on my system

That's a little bit longer than it takes on hour system (9Mbases take 19 hours on our Amazon cattle)

**Comment by lpoorvin** [comment]
The annotation seemed to finish fine.  Taking a quick glance at the output files, I see about what I expected.  One thing of note however is that the entire process completed much faster than I've seen with the previous build.  In the past, annotating a Pseudomonas with a genome size of ~6 Mbp would take at least 24 hours; this most current run with a similar genome finished in 4 hours (run on a dual Xeon E5-2643 v4 desktop).  It's great if the process is faster, I'm just a little concerned that some step was missed.  Has anyone else seen this level of speedup?

**Comment by thibaudnis** [comment]
Hi! Four hours does seem a bit fast to us too. Can you tell us the specs of your machine? 
Did you annotate the same genome with a previous release of PGAP, and were you able to compare the features produced (count, whether or not they have an informative gene description or a gene symbol)?
We'd also recommend running it again and checking if you get the same results, as for the run that took four hours.

**Comment by lpoorvin** [comment]
I was able to go back and re-annotate a different Pseudomonas genome, previously annotated with 2018-11-07.build3190.  The older build took >24 hrs to run, the current build completed in ~4.5 hrs.  Running a diff on the .gbk outputs showed almost no differences.  A few of the hypothetical gene products were listed differently, but that was about it.  One thing I've noticed, with the older builds, there were various steps that appeared to run with only one or two threads resulting in just a few processors getting fully utilized.  Every time I checked on it, the new build kept all 24 logical processors pinned at >95%, so I'm guessing the speedup is due to an increased number of threads running.

For reference, I'm running a desktop with dual Xeon E5-2643 v4 @ 3.4GHz (6 cores per CPU, running hyperthreaded for a total of 24 logical processors) with 64GB RAM.  The OS is Linux Mint 18.2 (based on Ubuntu Xenial 16.04 LTS) running on the 4.15.0-39-generic kernal.
